spec/03-plugins/08-datadog/01-log_spec.lua

```diff
@@ -168,4 +182,33 @@ describe("Plugin: datadog (log)", function()
     assert.contains("kong.datadog3_com.request.status.200:1|c|#T2:V2,T3:V3,T4", gauges)
     assert.contains("kong.datadog3_com.latency:%d+|g|#T2:V2:V3,T4", gauges, true)
   end)
+
+  it("should not return a runtime error (regression)", function()
+    local thread = threads.new({
+      function()
+        local socket = require "socket"
+        local server = assert(socket.udp())
+        server:settimeout(1)
+        server:setoption("reuseaddr", true)
+        server:setsockname("127.0.0.1", 9999)
+        local gauge = server:receive()
+        server:close()
+        return gauge
+      end
+    })
+    thread:start()
+
+    local res = assert(client:send {
+      method = "GET",
+      path = "/NonMatch",
+      headers = {
+        ["Host"] = "fakedns.com"
+      }
+    })
+
+    assert.res_status(404, res)
+    
+    local err_log = pl_file.read(helpers.test_conf.nginx_err_logs)
+    assert.is_nil(err_log:find("attempt to index field 'api' (a nil value)", nil, true))
```

Comment:
yeah, I comment my return code out and run the following test. It should fail, however, it succeeds.

```
  it("should not return a runtime error (regression)", function()
    local thread = threads.new({
      function()
        local socket = require "socket"
        local server = assert(socket.udp())
        server:settimeout(1)
        server:setoption("reuseaddr", true)
        server:setsockname("127.0.0.1", 9999)
        local gauge = server:receive()
        server:close()
        return gauge
      end
    })
    thread:start()

    local res = assert(client:send {
      method = "GET",
      path = "/NonMatch",
      headers = {
        ["Host"] = "fakedns.com"
      }
    })

    assert.res_status(404, res)
    
    local err_log = pl_file.read(helpers.test_conf.nginx_err_logs)
    print(err_log)
    assert.not_matches(err_log, "attempt to index field 'api' (a nil value)", nil, true)  
  end)
```

output
```
vagrant@vagrant-ubuntu-trusty-64:/kong$ bin/busted -v -o gtest spec/03-plugins/08-datadog/
[==========] Running tests from scanned files.
[----------] Global test environment setup.
[----------] Running tests from spec/03-plugins/08-datadog/01-log_spec.lua
[ RUN      ] spec/03-plugins/08-datadog/01-log_spec.lua @ 79: Plugin: datadog (log) logs metrics over UDP
[       OK ] spec/03-plugins/08-datadog/01-log_spec.lua @ 79: Plugin: datadog (log) logs metrics over UDP (256.73 ms)
[ RUN      ] spec/03-plugins/08-datadog/01-log_spec.lua @ 117: Plugin: datadog (log) logs only given metrics
[       OK ] spec/03-plugins/08-datadog/01-log_spec.lua @ 117: Plugin: datadog (log) logs only given metrics (93.25 ms)
[ RUN      ] spec/03-plugins/08-datadog/01-log_spec.lua @ 151: Plugin: datadog (log) logs metrics with tags
[       OK ] spec/03-plugins/08-datadog/01-log_spec.lua @ 151: Plugin: datadog (log) logs metrics with tags (66.23 ms)
[ RUN      ] spec/03-plugins/08-datadog/01-log_spec.lua @ 186: Plugin: datadog (log) should not return a runtime error (regression)
2017/07/19 16:30:59 [debug] 10410#0: [lua] globalpatches.lua:9: installing the globalpatches
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:73: load_plugins(): Discovering used plugins
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: syslog
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: ldap-auth
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: rate-limiting
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: correlation-id
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: jwt
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: request-termination
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: runscope
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: request-transformer
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: http-log
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: loggly
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: response-transformer
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: basic-auth
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: tcp-log
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: hmac-auth
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: oauth2
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: acl
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: bot-detection
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: udp-log
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: cors
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: file-log
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: ip-restriction
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: datadog
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: request-size-limiting
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: galileo
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: aws-lambda
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: statsd
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: response-ratelimiting
2017/07/19 16:30:59 [debug] 10410#0: [lua] kong.lua:101: load_plugins(): Loading plugin: key-auth
2017/07/19 16:30:59 [notice] 10410#0: using the "epoll" event method
2017/07/19 16:30:59 [notice] 10410#0: openresty/1.11.2.2
2017/07/19 16:30:59 [notice] 10410#0: built by gcc 4.6.3 (Ubuntu/Linaro 4.6.3-1ubuntu5)
2017/07/19 16:30:59 [notice] 10410#0: OS: Linux 3.13.0-123-generic
2017/07/19 16:30:59 [notice] 10410#0: getrlimit(RLIMIT_NOFILE): 65535:65535
2017/07/19 16:30:59 [notice] 10423#0: start worker processes
2017/07/19 16:30:59 [notice] 10423#0: start worker process 10425
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] globalpatches.lua:219: randomseed(): seeding PRNG from OpenSSL RAND_bytes()
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] globalpatches.lua:245: randomseed(): random seed: 135120151132 for worker nb 0
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "bot-detection": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "cors": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "jwt": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "oauth2": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "key-auth": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "ldap-auth": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "basic-auth": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "hmac-auth": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "ip-restriction": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "request-size-limiting": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "acl": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "rate-limiting": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "response-ratelimiting": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "request-transformer": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "response-transformer": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "aws-lambda": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "http-log": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "statsd": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "datadog": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "file-log": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "udp-log": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "request-termination": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "loggly": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "runscope": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "syslog": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "galileo": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "tcp-log": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *1 [lua] base_plugin.lua:12: init_worker(): executing plugin "correlation-id": init_worker
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] client.lua:483: lookup(): [dns-client] bad ipv4 query: '127.0.0.1' with type 33
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] client.lua:472: lookup(): [dns-client] ipv4 query: '127.0.0.1' with type 1
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:20: rewrite(): executing plugin "datadog": rewrite
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:24: access(): executing plugin "datadog": access
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] balancer.lua:33: cb(): fetching all upstreams
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] client.lua:531: lookup(): [dns-client] querying 'mockbin.com' for type 33 took 63.999891281128
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] client.lua:574: lookup(): [dns-client] querying 'mockbin.com' for type 33: entries 0 error nil nil
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] client.lua:531: lookup(): [dns-client] querying 'mockbin.com' for type 1 took 58.000087738037
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] client.lua:574: lookup(): [dns-client] querying 'mockbin.com' for type 1: entries 2 error nil nil
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:28: header_filter(): executing plugin "datadog": header_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:36: log(): executing plugin "datadog": log
2017/07/19 16:30:59 [debug] 10425#0: *6 [lua] statsd_logger.lua:62: counter(): [udp-log] sending data to statsd server: kong.datadog1_com.request.count:1|c
2017/07/19 16:30:59 [debug] 10425#0: *6 [lua] statsd_logger.lua:62: gauge(): [udp-log] sending data to statsd server: kong.datadog1_com.latency:245|g
2017/07/19 16:30:59 [debug] 10425#0: *6 [lua] statsd_logger.lua:62: gauge(): [udp-log] sending data to statsd server: kong.datadog1_com.request.size:101|g
2017/07/19 16:30:59 [debug] 10425#0: *6 [lua] statsd_logger.lua:62: counter(): [udp-log] sending data to statsd server: kong.datadog1_com.request.status.200:1|c
2017/07/19 16:30:59 [debug] 10425#0: *6 [lua] statsd_logger.lua:62: gauge(): [udp-log] sending data to statsd server: kong.datadog1_com.response.size:878|g
2017/07/19 16:30:59 [debug] 10425#0: *6 [lua] statsd_logger.lua:62: gauge(): [udp-log] sending data to statsd server: kong.datadog1_com.upstream_latency:98|g
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:20: rewrite(): executing plugin "datadog": rewrite
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:24: access(): executing plugin "datadog": access
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:28: header_filter(): executing plugin "datadog": header_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:36: log(): executing plugin "datadog": log
2017/07/19 16:30:59 [debug] 10425#0: *9 [lua] statsd_logger.lua:62: counter(): [udp-log] sending data to statsd server: kong.datadog2_com.request.count:1|c
2017/07/19 16:30:59 [debug] 10425#0: *9 [lua] statsd_logger.lua:62: counter(): [udp-log] sending data to statsd server: kong.datadog2_com.request.status.200:1|c
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:20: rewrite(): executing plugin "datadog": rewrite
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:24: access(): executing plugin "datadog": access
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:28: header_filter(): executing plugin "datadog": header_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:30:59 [debug] 10425#0: *2 [lua] base_plugin.lua:36: log(): executing plugin "datadog": log
2017/07/19 16:30:59 [debug] 10425#0: *11 [lua] statsd_logger.lua:62: counter(): [udp-log] sending data to statsd server: kong.datadog3_com.request.count:1|c|#T1:V1
2017/07/19 16:30:59 [debug] 10425#0: *11 [lua] statsd_logger.lua:62: counter(): [udp-log] sending data to statsd server: kong.datadog3_com.request.status.200:1|c|#T2:V2,T3:V3,T4
2017/07/19 16:30:59 [debug] 10425#0: *11 [lua] statsd_logger.lua:62: gauge(): [udp-log] sending data to statsd server: kong.datadog3_com.latency:58|g|#T2:V2:V3,T4
2017/07/19 16:31:00 [debug] 10425#0: *2 [lua] base_plugin.lua:20: rewrite(): executing plugin "datadog": rewrite
2017/07/19 16:31:00 [debug] 10425#0: *2 [lua] base_plugin.lua:28: header_filter(): executing plugin "datadog": header_filter
2017/07/19 16:31:00 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:31:00 [debug] 10425#0: *2 [lua] base_plugin.lua:32: body_filter(): executing plugin "datadog": body_filter
2017/07/19 16:31:00 [debug] 10425#0: *2 [lua] base_plugin.lua:36: log(): executing plugin "datadog": log
2017/07/19 16:31:00 [error] 10425#0: *13 lua entry thread aborted: runtime error: ./kong/plugins/datadog/handler.lua:67: attempt to index field 'api' (a nil value)
stack traceback:
coroutine 0:
	./kong/plugins/datadog/handler.lua: in function <./kong/plugins/datadog/handler.lua:52>, context: ngx.timer, client: 127.0.0.1, server: 0.0.0.0:9000

[       OK ] spec/03-plugins/08-datadog/01-log_spec.lua @ 186: Plugin: datadog (log) should not return a runtime error (regression) (9.41 ms)
[----------] 4 tests from spec/03-plugins/08-datadog/01-log_spec.lua (1864.43 ms total)

[----------] Global test environment teardown.
[==========] 4 tests from 1 test file ran. (1865.32 ms total)
[  PASSED  ] 4 tests.
```